src/odemis/driver/andorcam2.py

```diff
@@ -1872,426 +1990,355 @@ def _buffer_as_array(self, cbuffer, size, metadata=None):
-                    # kinetic should be approximately same as exposure + readout => play safe
-                    duration = max(accumulate, exposure + readout)
-                    logging.debug("Will get image every %g s (expected %g s)", accumulate, exposure + readout)
-                    need_reinit = False
-
-                # Acquire the images
-                self._start_acquisition()
-                tstart = time.time()
-                tend = tstart + duration
-                metadata = dict(self._metadata) # duplicate
-                metadata[model.MD_ACQ_DATE] = tstart
-                cbuffer = self._allocate_buffer(size)
-                array = self._buffer_as_array(cbuffer, size, metadata)
-
-                # first we wait ourselves the typical time (which might be very long)
-                # while detecting requests for stop
-                if self.acquire_must_stop.wait(max(0, duration - 0.1)):
-                    raise CancelledError()
-
-                # then wait a bounded time to ensure the image is acquired
-                try:
-                    while True:
-                        # cancelled by the user?
-                        if self.acquire_must_stop.is_set():
-                            raise CancelledError()
-
-                        # we actually _expect_ a timeout
+                        # This sometimes happen with 20024 (DRV_NO_NEW_DATA) or
+                        # 20067 (DRV_P2INVALID) on GetMostRecentImage16()
                         try:
-                            self.WaitForAcquisition(0.1)
-                        except AndorV2Error as ex:
-                            if ex.errno == 20024: # DRV_NO_NEW_DATA
-                                if time.time() > tend + 1:
-                                    logging.warning("Timeout after %g s", time.time() - tstart)
-                                    raise # seems actually serious
-                                else:
-                                    pass
+                            self.atcore.CancelWait()
+                            if self.GetStatus() == AndorV2DLL.DRV_ACQUIRING:
+                                self.atcore.AbortAcquisition()  # Need to stop acquisition to read temperature
+                            temp = self.GetTemperature()
+                        except AndorV2Error:
+                            temp = None
+                        # -999°C means the camera is gone
+                        if temp == -999:
+                            logging.error("Camera seems to have disappeared, will try to reinitialise it")
+                            self.Reinitialize()
                         else:
-                            break # new image!
+                            time.sleep(0.1)
+                            logging.warning("trying again to acquire image after error %s", ex)
+                        need_reconfig = True
+                        continue
```

Comment:
As this while loops are so extensive, you may add a comment here with which loop you continue.